Proxito: refactor canonical redirects (#10069)
readthedocs/readthedocs.org: readthedocs/proxito/README.rst

Read the Docs core team members can view the `Proxito design doc <https://github.com/readthedocs/el-proxito/blob/master/docs/design/architecture.rst>`_

> URL parts

Reviewer: This is a _really good_ start! 💯

Reviewer 2: Agreed, really happy to see this.

---------

> In our code we use the following terms to refer to the different parts of the URL:

Reviewer 2: ```suggestion
In proxito we use the following terms to refer to the different parts of the URL:
```

> url:
>    The full URL, for example ``https://docs.readthedocs.io/en/latest/index.html``.

Reviewer: ```suggestion
url:
   The full URL including the protocol, for example ``https://docs.readthedocs.io/en/latest/index.html``.
```

path:

>    The whole path from the URL, for example ``/en/latest/index.html``.

Reviewer 2: Does this include the GET args? What about `#fragment`

Author: This does not, I'll extend the definition to mention that.

domain:
   The domain/subdomain without the protocol, for example ``docs.readthedocs.io``.
language:
   The language of the documentation, for example ``en``.
version:
   The version of the documentation, for example ``latest``.
filename:

>    The name of the file being served, for example ``index.html``.

Reviewer: What if the whole path is `/section/page/myfile.html`? Would the `filename` be just `myfile.html` or it will include the rest as well?

> path prefix:
>    The path prefix of the URL without version or language,
>    for a normal project this is ``/``, and for subprojects this is ``/projects/<subproject-alias>/``.

Reviewer: ```suggestion
path prefix:
   The path prefix of the URL without version or language,
   for a normal project this is ``/``, and for subprojects this is ``/projects/<subproject-alias>/``.
   Note this could be different for project defining their own urlconf.
```

.. code:: text

                         URL
   |------------------------------------------------|
                                        path
                              |---------------------|
    https://docs.readthedocs.io/en/latest/index.html
   |-------|-------------------|--|------|----------|
    protocol         |          |     |       |
                   domain       |     |       |
                             language |       |
                                    version   |

>                                            filename

Reviewer: :100:

Reviewer 2: This is great. Though this is varying a bit from urlparse in some ways which make sense, but we should be explicit if we don't care about the other parts (eg. `query` & `fragment`).

`urlparse("scheme://netloc/path;parameters?query#fragment")`

https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse
